components/org.wso2.carbon.identity.application.authenticator.fido2/src/test/java/org/wso2/carbon/identity/application/authenticator/fido2/core/WebAuthnServiceTest.java

```diff
@@ -0,0 +1,663 @@
+        when(FIDOUtil.writeJson(any(AssertionRequestWrapper.class))).thenReturn("assertionRequest");
+
+        String response = webAuthnService.startUsernamelessAuthentication(ORIGIN);
+        Assert.assertEquals(response, "assertionRequest");
+    }
+
+    @Test(description = "Test case for finishAuthentication() method when assertion fails", expectedExceptions = {
+            AuthenticationFailedException.class
+    }, priority = 7)
+    public void testFinishAuthenticationFailedAssertion() throws JsonProcessingException,
+            MalformedURLException, AuthenticationFailedException, AssertionFailedException {
+
+        when(objectMapperMock.readValue(finishAuthenticationResponseString, AssertionResponse.class))
+                .thenReturn(finishAuthenticationResponse);
+        when(fido2Cache.getValueFromCacheByRequestId(any(FIDO2CacheKey.class))).thenReturn(fido2CacheEntry);
+        when(fido2CacheEntry.getAssertionRequest()).thenReturn("assertionRequest");
+        when(fido2CacheEntry.getOrigin()).thenReturn(new URL(ORIGIN));
+        when(objectMapperMock.readValue("assertionRequest", AssertionRequest.class)).thenReturn(assertionRequest);
+        when(relyingParty.finishAssertion(any(FinishAssertionOptions.class))).thenReturn(assertionResult);
+        when(assertionResult.isSuccess()).thenReturn(false);
+        when(assertionResult.getUsername()).thenReturn(USERNAME);
+        when(assertionResult.getWarnings()).thenReturn(new ArrayList<>());
+
+        webAuthnService.finishAuthentication(USERNAME, TENANT_DOMAIN, USER_STORE_DOMAIN,
+                finishAuthenticationResponseString);
+    }
+
+    @Test(description = "Test case for finishAuthentication() method when assertion success", priority = 8)
+    public void testFinishAuthenticationSuccessfulAssertion() throws JsonProcessingException,
+            MalformedURLException, AuthenticationFailedException, AssertionFailedException {
+
+        when(objectMapperMock.readValue(finishAuthenticationResponseString, AssertionResponse.class))
+                .thenReturn(finishAuthenticationResponse);
+        when(fido2Cache.getValueFromCacheByRequestId(any(FIDO2CacheKey.class))).thenReturn(fido2CacheEntry);
+        when(fido2CacheEntry.getAssertionRequest()).thenReturn("assertionRequest");
+        when(fido2CacheEntry.getOrigin()).thenReturn(new URL(ORIGIN));
+        when(objectMapperMock.readValue("assertionRequest", AssertionRequest.class)).thenReturn(assertionRequest);
+        when(relyingParty.finishAssertion(any(FinishAssertionOptions.class))).thenReturn(assertionResult);
+        when(assertionResult.isSuccess()).thenReturn(true);
+        when(assertionResult.getUsername()).thenReturn(USERNAME);
+        when(assertionResult.getWarnings()).thenReturn(new ArrayList<>());
+
+        webAuthnService.finishAuthentication(USERNAME, TENANT_DOMAIN, USER_STORE_DOMAIN,
```

Comment:
Don't we need an assertion for this?

Reply:
We cannot have a assertion here as there's no return value. Expectation is it should NOT throw any exception.